### PR TITLE
[#1212] HTTPS inside Kubernetes

### PR DESCRIPTION
**Pull request by ljupcovangelski**
- Added option for helm to deploy with letsencrypt.
- After creation, the ingress controller reads the `host` entries in the `ingress` resources and creates the letsencrypt certificates.
- Letsencrypt doesn't work with the loadbalancer url, because the url is very long.

Potential usage:
- Run `airy create`
- Add DNS entry for `myairy.mydomain.com` -> created loadbalancer
- Patch the ingress resources
```
for ingress in $(kubectl get ingress --no-headers -o custom-columns=":metadata.name"); do kubectl get ingress ${ingress} -o json |  jq '(.spec.rules[0].host="myairy.mydomain.com")' | kubectl apply -f -; done
```

And the letsencrypt certificates are created. Also redirect from https to http is configured.

### CODE REVIEW

File: cli/pkg/providers/aws/aws.go [hunk 1]
```diff
@@ -49,42 +49,45 @@ func (p *provider) GetOverrides() tmpl.Variables {
 	}
 }
 
-func (p *provider) PostInstallation(dir workspace.ConfigDir) error {
-	conf, err := dir.LoadAiryYaml()
-	if err != nil {
-		return err
-	}
+func (p *provider) PostInstallation(providerConfig map[string]string, dir workspace.ConfigDir) error {
+	if providerConfig["hostUpdate"] != "false" {
```
Comment (ljupcovangelski) [inline cli/pkg/providers/aws/aws.go:53]: We control with a flag whether to update the hosts after the installation or `global.kubernetes.host` is used in the ingress definitions and in the `hostnames` configMap.

File: cli/pkg/workspace/template/src/airy.yaml [hunk 2]
```diff
@@ -4,14 +4,20 @@ kubernetes:
...
     {{ $k }}: {{ $v }}
   {{- end }}
-{{- end }}
+  {{- end }}
 security:
   allowedOrigins: "*"
   jwtSecret: {{ randAlphaNum 128 }}
+components:
+  sources:
+  integration:
```
Comment (ljupcovangelski) [inline cli/pkg/workspace/template/src/airy.yaml:23]: I thought it also makes sense to add these items when creating an skel `airy.yaml` with `--init-only`

File: docs/docs/getting-started/installation/aws.md [hunk 3]
```diff
@@ -225,6 +229,56 @@ At this point, the frontend and the API services of `Airy Core` should be access
 airy api endpoint
 ```
 
+### HTTPS using Let's Encrypt
```
Comment (ljupcovangelski) [inline docs/docs/getting-started/installation/aws.md:232]: @steffh please have a look at this section.

File: docs/docs/getting-started/installation/aws.md [hunk 4]
```diff
@@ -225,6 +229,56 @@ At this point, the frontend and the API services of `Airy Core` should be access
...
+  letsencryptEmail: "mymail@myhostname.com"
+
+```
+
+The `kubernets.host` value should be set to your desired hostname. Configure the e-mail address you want to use for your Let's Encrypt registration under `ingress.letsencryptEmail`.
+
+After setting these parameters, create your `Airy Core` instance with the following option:
+
+```sh
+airy create --provider aws --provider-config hostUpdate=false
```
Comment (ljupcovangelski) [inline docs/docs/getting-started/installation/aws.md:260]: @steffh if the `--provider-config` is problematic (we use it also to overwrite vpcId and instanceType for AWS), I will try to remove it.

File: infrastructure/helm-chart/charts/ingres/templates/ingress-controller-https-letsencrypt.yaml [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+apiVersion: apps/v1
+kind: StatefulSet
+metadata:
+  name: traefik-ingress-controller
+  namespace: kube-system
+  labels:
+    k8s-app: traefik-ingress-lb
+spec:
+  serviceName: traefik-ingress-lb
+  replicas: 0
```
Comment (ljupcovangelski) [inline infrastructure/helm-chart/charts/ingres/templates/ingress-controller-https-letsencrypt.yaml:62]: Started with 0 replicas and instructions how to scale up after the DNS is configured, because Let's Encrypt will return error before that and might block/throttle requests in the next hour.

File: infrastructure/helm-chart/templates/ingress-ui.yaml [hunk 6]
```diff
@@ -8,7 +8,7 @@ metadata:
     traefik.frontend.rule.type: PathPrefixStrip
 spec:
   rules:
-    - host: {{ get (urlParse .Values.global.kubernetes.host) "host" }}
+    - host: {{ .Values.global.kubernetes.host }}
```
Comment (ljupcovangelski) [inline infrastructure/helm-chart/templates/ingress-ui.yaml:11]: In the `airy.yaml` file now we specify only the hostname and the protocol (http or https) is specified in `ingress.https` (true or false).

File: infrastructure/helm-chart/charts/ingres/Chart.yaml [hunk 7]
```diff
@@ -0,0 +1,5 @@
+apiVersion: v1
+appVersion: "1.0"
+description: A Helm chart for Airy Platform networking resources
+name: ingress
+version: 1.0
```
Comment (ljupcovangelski) [inline infrastructure/helm-chart/charts/ingres/Chart.yaml:5]: Eventually the `Ingress controller` chart should be separate from `Airy Core`, so that people can customize how and if they want to deploy an ingress controller.

**Review by chrismatix — DISMISSED** [review]

File: docs/docs/getting-started/installation/aws.md [hunk 8]
```diff
@@ -225,6 +229,56 @@ At this point, the frontend and the API services of `Airy Core` should be access
...
+You should create a CNAME DNS record for the hostname that you set under `kubernetes.host` in the previous step to point to the hostname of the LoadBalancer, created by AWS for the ingress service:
+
+```sh
+export KUBECONFIG="PATH/TO/DIR/kube.conf"
+kubectl get --namespace kube-system service traefik --output jsonpath='{.status.loadBalancer.ingress[0].hostname}{"\n"}'
+```
+
+#### Start the ingress controller
+
+If the ingress controller is started before the DNS record is added, the Let's Encrypt servers will block an throttle the registration attempts. That is why we recommend starting the ingress controller after the DNS record is added.
```
Comment (pascal-airy) [inline docs/docs/getting-started/installation/aws.md:274]:
```suggestion
If the ingress controller is started before the DNS record is added, the Let's Encrypt servers will block and throttle the registration attempts. That is why we recommend starting the ingress controller after the DNS record is added.
```

File: infrastructure/helm-chart/templates/configmap.yaml [hunk 9]
```diff
@@ -4,7 +4,15 @@ metadata:
...
 data:
-  HOST: {{ .Values.global.kubernetes.host }}
+  {{ if .Values.global.ingress }}
+   {{ if .Values.global.ingress.https }}
+  HOST: https://{{ .Values.global.kubernetes.host }}
+   {{ else }}
+  HOST: http://{{ .Values.global.kubernetes.host }}
+   {{ end }}
+  {{ else }}
+  HOST: http://{{ .Values.global.kubernetes.host }}
```
Comment (pascal-airy) [inline infrastructure/helm-chart/templates/configmap.yaml:14]: why is this line here twice?

Comment (ljupcovangelski) [reply infrastructure/helm-chart/templates/configmap.yaml:14]: The logic is:
```
if the ingress section exists
  if ingress.https == true
    host: https...
  else (ingress.https != true)
    host: http...
else (ingress section doesn't exist)
  host: http...
```

Comment (pascal-airy) [reply infrastructure/helm-chart/templates/configmap.yaml:14]: ah because the evaluation of `if .Values.global.ingress.https` would fail if `if .Values.global.ingress` isn't there

**Review by pascal-airy — APPROVED** [review]
👍
